### DISCUSSION TITLE
NormalModule original source is undefined when cache is enabled

### DISCUSSION THREAD

**Discussion by yoriiis** [webpack/webpack#16735] [Q&A]
# Bug report

**What is the current behavior?**

When the webpack cache is enabled, the module source inside the compilation is not accessible, it returns _randomly_ `undefined`. The source is get with [`NormalModule.originalSource()._value`](https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/master/src/index.ts#L159) in the plugin `svg-chunk-webpack-plugin`.

The plugin uses the `processAssets` hook with the stage `PROCESS_ASSETS_STAGE_ADDITIONAL`. All stages listed in the webpack documentation produces the same error ([list of asset processing stages](https://webpack.js.org/api/compilation-hooks/#list-of-asset-processing-stages)).

The cache is enabled with this code

```js
cache: {
	type: 'filesystem'
}
```

The error log produced by webpack is the following. The error appears when the module source is decoded from JSON.

```txt
[webpack-cli] HookWebpackError: Unexpected token u in JSON at position 0
    at makeWebpackError (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/webpack/lib/HookWebpackError.js:48:9)
    at /Users/yoriiis/svg-chunk-webpack-plugin/node_modules/webpack/lib/Compilation.js:3060:12
    at eval (eval at create (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:20:1)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
-- inner error --
SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at SvgSprite.optimizeSvg (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:41:24)
    at /Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:120:52
    at Array.map (<anonymous>)
    at SvgSprite.processEntry (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:120:21)
    at /Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:97:68
    at Array.map (<anonymous>)
    at SvgSprite.addAssets (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:97:32)
    at fn (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/webpack/lib/Compilation.js:536:19)
    at _next0 (eval at create (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:12:17)
caused by plugins in Compilation.hooks.processAssets
SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at SvgSprite.optimizeSvg (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:41:24)
    at /Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:120:52
    at Array.map (<anonymous>)
    at SvgSprite.processEntry (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:120:21)
    at /Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:97:68
    at Array.map (<anonymous>)
    at SvgSprite.addAssets (/Users/yoriiis/svg-chunk-webpack-plugin/lib/index.js:97:32)
    at fn (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/webpack/lib/Compilation.js:536:19)
    at _next0 (eval at create (/Users/yoriiis/svg-chunk-webpack-plugin/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:12:17)
```

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Clone the code from https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8
2. Execute `nvm use` or switch to Node.js `v12.22.7`
3. Install dependencies with `npm install`
4. Run the example in development mode with `npm run start:example`
5. The webpack cache is created in `node_modules/.cache/webpack`
6. Exit the watcher and re-run it
7. The error appears when the plugin tried to access the module source (svg file)
8. Remove the webpack cache
9. Re-run the watcher
10. It works, because the cache is empty

**What is the expected behavior?**

The module source should be accessible.

**Other relevant information:**
webpack version: `5.72.0` (already tried with the latest without success)
Node.js version: `12.22.7`
Operating System: `macOS 12.6`
Additional tools: -

**Comment by alexander-akait** [comment]
Why do you use `NormalModule.originalSource()._value`, it is private and unsafe to use (you can't use the field), what do you need? Do you get this in the loader or the plugin? There is https://github.com/webpack-contrib/html-minimizer-webpack-plugin/blob/master/src/index.js#L265 example of how to cache and parallels handling assets, you can get a code, feel free to feedback

**Reply by yoriiis** [reply]
Hello @alexander-akait, thanks for the response, I need the original source of the file and I get this from the plugin.
I couldn't find any documentation so I searched the compilation object for what might work. Do you have a better way to get this value?

About the cache, the plugin doesn't manage anything about it, I don't know if it's necessary? A project that consumes the plugin may have caching enabled and it is this behavior that is causing the problem.

**Reply by yoriiis** [reply]
@alexander-akait have you any updates about this please? Thanks

**Reply by alexander-akait** [reply]
Sorry for delay, but webpack tries to save memory (especial when cache enabled), so at some moment they do not have content and we store already processed files, i.e.
the first run - original content -> transform/generate -> cache results
the second run - get from cache and use it (and here you can have a problem)
the third run is the same as the second

So my qeustion why do not implement cache and avoid using `private` properties?

There is a potential place to cache https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/master/src/index.ts#L158, you can use logic from here - https://github.com/webpack-contrib/html-minimizer-webpack-plugin/blob/master/src/index.js#L398

Anyway I can help to implement this if you need, feel free to feedback

**Reply by yoriiis** [reply]
I'll be happy to code this on the plugin but I need help as it's a bit complex and undocumented. (I studied your example, but I don't understand how I can apply it to my plugin).

The plugin needs to get the source of assets **by entrypoint**, so I used the [getSvgsDependenciesByEntrypoint](https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/master/src/index.ts#L173) function to get those assets. More detail about this function in this issue with the webpack core team: https://github.com/webpack/webpack/issues/12202#issuecomment-745537821.

I want to replace the private properties, but what instead? (It must indeed be compatible with cache enabled or disabled).

My idea of the build steps:
- Get svgs sources (from cache or not)
- Generate svgs sprite
- Store svgs sprite in the cache (to improve build performance)

---

How do you want us to code this part? Feel free to contact me

**Reply by yoriiis** [reply]
@alexander-akait I found an alternative to get the `NormalModule` source, tell me what you think.

**Before**

```js
NormalModule.originalSource()._value
```

**After**

```js
compilation.codeGenerationResults.get(NormalModule).sources.get('javascript').source()
```

Changes are available on [svg-chunk-webpack-plugin#8](https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8/files).

---

Regarding the cache, I think it can improve the performance of the plugin, which is currently a bit slow. All SVG sprites are rebuilt on every change even if there is no need.

Ideally, I'd like to store the sprite in cache and only generate it when needed. How could i do that?

**Reply by alexander-akait** [reply]
Sorry for delay 

> Ideally, I'd like to store the sprite in cache and only generate it when needed. How could i do that?

Yeah, for one file we use https://github.com/webpack-contrib/json-minimizer-webpack-plugin/blob/master/src/index.js#L122, but for multiple fies you can use `cache.getLazyHashedEtag(source)` for each source (i.e. images) which you have and then merge etags using `cache.mergeEtags(etag1, etag2)`, then yse `cache.getItemCache(name, eTag)` (`name` is name of your ouput file, should not be absolute) to get cached item, if not found run your logic for generation and use `cacheItem.storePromise`

Just take a look at the `html-minimizer-webpack-plugin` plugin and implement the same logic only for multiple input images

Feel free to feedback if you need help

**Reply by yoriiis** [reply]
@alexander-akait I can't get data from cache, the value of `output` is always `undefined`, any idea?

```js
const cache = compilation.getCache('SvgChunkWebpackPlugin')
const source = new RawSource(sprite); // `sprite` contains the source of a SVG sprite
const eTag = cache.getLazyHashedEtag(source);
const cacheItem = cache.getItemCache('sprites/home.svg', eTag);
const output = await cacheItem.getPromise();

if (!output) {
    await cacheItem.storePromise({
        source
    });
    compilation.emitAsset('sprites/home.svg', source);
}
```

**Reply by alexander-akait** [reply]
`const source = new RawSource(sprite);` you should create a source when you don't have cache, how you get images for sprites?

**Reply by yoriiis** [reply]
@alexander-akait But to use the cache, one of the parameters is the `source` (with `getLazyHashedEtag`) and this needs to be "calculated" in the plugin.

The source of each svg sprite is generated like this (on every compilation):
- Get all SVG dependencies for each chunks
- Get svgs sources
- Merge svgs sources to create svg sprite (by chunk)
- Sprite source are now available

**Reply by alexander-akait** [reply]
> Get svgs sources

Can you show me a code how you do it, because you should have `source`s here already

**Reply by yoriiis** [reply]
The loader exports the content of the SVG asset as a JSON string. Then, the plugin get the source of each SVG with the code below:

```js
JSON.parse(compilation.codeGenerationResults.get(normalModule).sources.get('javascript').source())
```

But, the sprite is composed by the plugin, it is the merge of severals SVG according to the entrypoint dependencies.

---

The following entries

```js
// entry-1.js
import 'svg-a.svg'
import 'svg-b.svg'
import 'svg-c.svg'
```

```js
// entry-2.js
import 'svg-a.svg'
import 'svg-d.svg'
import 'svg-e.svg'
```

Will produce

```html
<!-- entry-1.svg -->
<svg>
    <symbol id="svg-a"></symbol>
    <symbol id="svg-b"></symbol>
    <symbol id="svg-c"></symbol>
</svg>
```

```html
<!-- entry-2.svg -->
<svg>
    <symbol id="svg-a"></symbol>
    <symbol id="svg-d"></symbol>
    <symbol id="svg-e"></symbol>
</svg>
```

**Reply by alexander-akait** [reply]
I think it should work (pseudo code):
```js
const spriteName = 'entry.svg';
const cache = compilation.getCache('SvgChunkWebpackPlugin');

/* Note - order is matter, because we need to generate a sprite in the order in which we have imports */
const files = [compilation.codeGenerationResults.get(normalModule).sources.get('javascript').source(), /* ... other */];
/* Calculate eTag for the each file in sprite */
const eTags = files.map((file) => cache.getLazyHashedEtag(file));

/* Merge them into one eTag */
let eTag;

for (var i in arr) {
  eTag = cache.mergeEtags(eTag, arr[i]);
}

/* Get from cache */
const cacheItem = cache.getItemCache(spriteName, eTag);

let output = await cacheItem.getPromise();

if (!output) {
    // LOGIC TO GENERATE SPRITE, when we don't have an item in cache
    // https://github.com/webpack/webpack/blob/main/declarations.d.ts#L259
    output = { source: myFunctionToGenerateSpriteAndReturnSource() };

    /* Store cache */
    await cacheItem.storePromise({
        source: output.source,
        // myAnyKey: myAnyValue, if you have dynamic generated filename, you can store any kind of information here
    });
}

 compilation.emitAsset(spriteName, output.source, { /* Assets info */ });
```

**Reply by alexander-akait** [reply]
Yeah, your logic can be more complex, but I think the above code will help to embed it into the logic

**Reply by yoriiis** [reply]
@alexander-akait SVG files are marked as a side effect by the loader (see [loader.ts#L14](https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/master/src/loader.ts#L14)) and with the cache it appears in production mode that the source of the file is not available from the second build.

I think it's linked to https://github.com/webpack/webpack/issues/15214. If the loader returns the content like in the following example, webpack still keeps the module but it's also included in the JavaScript file, which I don't want.

```js
returns `module.exports = ${content}`
```

How can I correct this?

**Reply by alexander-akait** [reply]
Hm, sorry, I don't undestand, `side effect` should not be a problem for the cache (the linked issue was just a bug in our logic and it was fixed), do you mean you don't have files from the loader (i.e. sprites) on the second run, because the loader did start due cache, right?

**Reply by yoriiis** [reply]
No worries @alexander-akait.

On https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8, the webpack cache is enabled and I have a diff between the first build and the following.

I added a log in [getSvgsDependenciesByEntrypoint](https://github.com/yoriiis/svg-chunk-webpack-plugin/blob/cache-issue/src/index.ts#L184) and the following command produced the screenshot below.

```bash
rm -rf node_modules/.cache/webpack
npm run build:example
npm run build:example
```

![image](https://user-images.githubusercontent.com/2563298/229744973-1deec162-0bb1-4b60-ac58-492c10fb3402.png)

The first build found NormalModule dependencies, but the following not. It appears only when the cache is enabled.

---

I talked about side effect because we had the discussion in https://github.com/webpack/webpack/issues/12202#issuecomment-745537821 and it seems also related. Sorry if it's confusing.

**Reply by alexander-akait** [reply]
> The first build found NormalModule dependencies, but the following not. It appears only when the cache is enabled.

That is expected because on the second run webpack doesn't run the loader, otherwise why we need cache? :smile: 

So the solution - calculate `eTag` in a place where you get sources and keep the hash of the previous build, so before run you will have `eTag = undefined`, after first run you will have `eTag = "ab12cd34"`, on the second run because `eTag` was not changed (because you don't have sources and nothing was changed) you just reemit cached `const cacheItem = cache.getItemCache(spriteName, eTag);` (sprite name you can get from options, and eTag from previous build)

Also you always have `hash` of build so you can undestand that nothing was changed, can you try to implement a basic cache logic (on the first run) and I can help you to implement a cache logic for the second/third/etc runs (if you stuch with it)

**Reply by yoriiis** [reply]
@alexander-akait Thanks for the explanations. I tried to implement it but it's not convincing and the current workflow is not efficient (I have several open questions and errors).

Maybe you could do a PR for this implementation?

**Reply by alexander-akait** [reply]
Oh, I can try to plan it, but I really don't have many times, will be great if you will try it and if you will have not a success I will finish

**Reply by yoriiis** [reply]
@alexander-akait Got it, I'll try to set it up and show you a version soon.

In your explanation above, `eTag` is a string (a hash) but when it is generated by `getLazyHashedEtag` or `mergeEtags` it is a object. I tried to store it in the cache but webpack trigger a serialization error. How do I handle it?

**Reply by alexander-akait** [reply]
Example of usage:
```
const eTag = cache.getLazyHashedEtag(source);
const cacheItem = cache.getItemCache(name, eTag);
```

`mergeEtags` and `getLazyHashedEtag` return the same `Etag` - https://github.com/webpack/webpack/blob/main/types.d.ts#L614

So you can use any, but in your case you need to merge tags because you have multiple source objects

**Reply by yoriiis** [reply]
@alexander-akait Hum, sorry, I don't understand. How you keep `eTag` from previous build?

**Reply by alexander-akait** [reply]
`class MyClass { #prevETag; }` and just write it in a method where you collect sources and calculate ETag :smile: 

So when a loader was not run and you don't have sources (there are different solutions how to undestand it - check previously hash of build, but it is not effective or check changed files https://stackoverflow.com/questions/43140501/can-webpack-report-which-file-triggered-a-compilation-in-watch-mode, and more other solutions) you just use `prevETag` and reemit asset from cache

**Reply by yoriiis** [reply]
@alexander-akait Yes :) But in build production mode, it's a new instance of the plugin that is created each time so if I store the value in a class property, it will always be `undefined`. I am wrong ?

The `eTag` value should be store is the cache also I imagine 🤔

**Reply by alexander-akait** [reply]
> @alexander-akait Yes :) But in build production mode, it's a new instance of the plugin that is created each time so if I store the value in a class property, it will always be undefined. I am wrong ?

We are creating a plugin instance only one and run `apply` method

**Reply by yoriiis** [reply]
@alexander-akait Okay that's why I didn't understand 😄
I just tested and the `constructor` is called at each build, am I missing something? 

![image](https://user-images.githubusercontent.com/2563298/230035656-692e6b4c-5e32-42b0-917f-7f8408e5d0b7.png)

**Reply by alexander-akait** [reply]
Oh, I see what do you mean, in this case - yes, you can use webpack cache to cache such things too and restore them from cache, you can use options for plugin as a key

**Reply by yoriiis** [reply]
> you can use options for plugin as a key

What do you mean?

---

@alexander-akait I've implemented a early version of the cache feature in https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8, could you take a look?

I having some issues with cache invalidation (i.e. force the process of sprite and cache generation).
Below an example of compilation hash saved across build.

```txt
# build 1
previousCompilationHash undefined
currentCompilationHash daa0672f163c8b20f4f7
// create cache

# build 2 (no changes) => Why hash are different?
previousCompilationHash daa0672f163c8b20f4f7
currentCompilationHash fb295c87252ad7de3ea2
// use cache

# build 3 (no changes)
previousCompilationHash fb295c87252ad7de3ea2
currentCompilationHash fb295c87252ad7de3ea2
// use cache

# build 4 (with changes) => Need to invalidate cache
previousCompilationHash fb295c87252ad7de3ea2
currentCompilationHash 0a7635d67ca3e369e154
// use cache

# build 5 (no changes) Why hash are different?
previousCompilationHash 0a7635d67ca3e369e154
currentCompilationHash 3f3578296be1916afcf8
// use cache

# build 5 (no changes)
previousCompilationHash 3f3578296be1916afcf8
currentCompilationHash 3f3578296be1916afcf8
// use cache
```

**Reply by yoriiis** [reply]
@alexander-akait I'm adding some info about this problem (runs are done in production mode).

My idea was that if there is no cache available, it is the first run, so I'll generate all the files and update the cache.

On the second run:
- if the compilation hash is the same as the previous one, there is no change so I use the cache
- if the compilation hash is different, there are changes and I do the sprite generation and cache update process

But, the 2nd run generates a different hash while there is no change, the logic would be to do the sprite generation and cache update process, but the compilation does not return any modules, which seems normal to me since there were no changes.

1. Why does the hash change on the 2nd compilation? I don't see how to detect if I should use cache or not based on file changes (and hash). I don't reproduce this problem in development mode, nor with the cache disabled.
2. With the cache enabled, webpack loader seems not to fire when there are changes, why? It can generate inaccurate files.

**Reply by alexander-akait** [reply]
> But, the 2nd run generates a different hash while there is no change, the logic would be to do the sprite generation and cache update process, but the compilation does not return any modules, which seems normal to me since there were no changes.

Yeah, that is why we should use only eTag from sources... I need to look at your code (hope will find time soon), anyway without looking at code I have two solutions here:
1. Get changed files and undestand you have or not have changed files which used in sprite
2. Another way use modules as cache and use WeakMaps (to avoid using extra memory) to when module(s) will changed you will have other ETag and it run sprite compilation and emit a new asset

> With the cache enabled, webpack loader seems not to fire when there are changes, why? It can generate inaccurate files.

Yeah, because you have cached files, so you don't need to spend time to run code again

**Reply by alexander-akait** [reply]
@yoriiis I rewrote your plugin using cache and fix memory leaking, it is not typescript (because types require a little more time), but you can adopt and rewrite it to typescript, code:
```js
'use strict';
/**
 * @license MIT
 * @name SvgChunkWebpackPlugin
 * @version 2.0.1
 * @author: Yoriiis aka Joris DANIEL <joris.daniel@gmail.com>
 * @copyright 2021 Joris DANIEL
 **/
const path = require('path');
const svgstore = require('svgstore');
const { optimize } = require('svgo');
const extend = require('extend');
const templatePreview = require('./preview');
const PACKAGE_NAME = require('../package.json').name;
const REGEXP_NAME = /\[name\]/i;
const REGEXP_HASH = /\[hash\]/i;
const REGEXP_CHUNKHASH = /\[chunkhash\]/i;
const REGEXP_CONTENTHASH = /\[contenthash\]/i;

class SvgSprite {
	/**
	 * Instanciate the constructor
	 * @param {options}
	 */
	constructor(options = {}) {
		// This need to find plugin from loader context
		this.PLUGIN_NAME = PACKAGE_NAME;
		/**
		 * Optimize SVG with Svgo
		 * @param {String} filepath SVG filepath from Webpack compilation
		 * @returns {Promise<Svgs>} Svg name and optimized content with Svgo
		 */
		this.optimizeSvg = async (compilation, moduleDependency) => {
			const source = JSON.parse(
				compilation.codeGenerationResults
					.get(moduleDependency)
					.sources.get('javascript')
					.source()
			);
			const svgOptimized = await optimize(source, Object.assign({}, this.options.svgoConfig));

			return {
				name: path.basename(moduleDependency.userRequest, '.svg'),
				content: svgOptimized.data
			};
		};
		const DEFAULTS = {
			filename: '[name].svg',
			svgstoreConfig: {
				cleanDefs: false,
				cleanSymbols: false,
				inline: true
			},
			svgoConfig: {},
			generateSpritesManifest: false,
			generateSpritesPreview: false
		};
		this.options = extend(true, DEFAULTS, options);
		this.hookCallback = this.hookCallback.bind(this);
		this.addAssets = this.addAssets.bind(this);
	}
	/**
	 * Apply function is automatically called by the Webpack main compiler
	 * @param {Object} compiler The Webpack compiler variable
	 */
	apply(compiler) {
		compiler.hooks.thisCompilation.tap('SvgSprite', this.hookCallback);
	}
	/**
	 * Hook expose by the Webpack compiler
	 * @param {Object} compilation The Webpack compilation variable
	 */
	async hookCallback(compilation) {
		// PROCESS_ASSETS_STAGE_ADDITIONAL: Add additional assets to the compilation
		compilation.hooks.processAssets.tapPromise(
			{
				name: 'SvgChunkWebpackPlugin',
				// For better compatibility with future webpack versions
				stage: compilation.compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
			},
			async () => {
				await this.addAssets(compilation);
			}
		);
	}
	/**
	 * Add assets
	 * The hook is triggered by webpack
	 */
	async addAssets(compilation) {
		// For better compatibility with future webpack versions
		const RawSource = compilation.compiler.webpack.sources.RawSource;
		// We **should not** store such values using class properties, because it is a memory leak
		// We need allow GC to clean up the memory
		const sprites = [];
		const spritesManifest = {};
		const spritesList = [];

		const cache = compilation.getCache('SvgChunkWebpackPlugin');

		await Promise.all(
			Array.from(compilation.entrypoints.keys()).map(async (entryName) => {
				let svgsDependencies = [];

				compilation.entrypoints.get(entryName).chunks.forEach((chunk) => {
					for (const module of compilation.chunkGraph.getChunkModulesIterable(chunk)) {
						if (module.buildInfo && module.buildInfo.SVG_CHUNK_WEBPACK_PLUGIN) {
							svgsDependencies.push(module);
						}
					}
				});

				// Need to sort (**always**), to have deterministic build
				svgsDependencies = svgsDependencies.sort((a, b) => {
					if (a.name < b.name) {
						return -1;
					}

					if (a.name > b.name) {
						return 1;
					}

					return 0;
				});

				const eTag = svgsDependencies
					.map((item) => {
						return cache.getLazyHashedEtag(item._source);
					})
					.reduce((result, item) => {
						return cache.mergeEtags(result, item);
					});
				const cacheItem = cache.getItemCache(entryName, eTag);

				let output = await cacheItem.getPromise();

				if (!output) {
					const svgsOptimized = await Promise.all(
						svgsDependencies.map((moduleDependency) =>
							this.optimizeSvg(compilation, moduleDependency)
						)
					);

					const sprite = this.generateSprite(svgsOptimized);
					const source = new RawSource(sprite, false);
					const manifestItem = svgsDependencies.map((moduleDependency) =>
						path.relative(compilation.options.context, moduleDependency.userRequest)
					);
					const spriteItem = {
						name: entryName,
						content: source,
						svgs: svgsDependencies.map((moduleDependency) =>
							path.basename(moduleDependency.userRequest, '.svg')
						)
					};

					output = {
						entryName,
						source,
						filename: this.getFileName({ entryName, sprite }),
						manifestItem,
						spriteItem
					};

					await cacheItem.storePromise(output);
				}

				compilation.emitAsset(output.filename, output.source);

				sprites.push({ entryName, source: output.source });
				spritesManifest[output.entryName] = output.manifestItem;
				spritesList.push(output.spriteItem);
			})
		);

		if (this.options.generateSpritesManifest || this.options.generateSpritesPreview) {
			// Need to sort (**always**), to have deterministic build
			const eTag = sprites
				.sort((a, b) => {
					if (a.entryName < b.entryName) {
						return -1;
					}

					if (a.entryName > b.entryName) {
						return 1;
					}

					return 0;
				})
				.map((item) => {
					return cache.getLazyHashedEtag(item.source);
				})
				.reduce((result, item) => {
					return cache.mergeEtags(result, item);
				});

			if (this.options.generateSpritesManifest) {
				const cacheItem = cache.getItemCache('sprites-manifest.json', eTag);

				let output = await cacheItem.getPromise();

				if (!output) {
					output = new RawSource(JSON.stringify(spritesManifest, null, 2), false);

					await cacheItem.storePromise(output);
				}

				compilation.emitAsset('sprites-manifest.json', output);
			}

			if (this.options.generateSpritesPreview) {
				const cacheItem = cache.getItemCache('sprites-preview.html', eTag);

				let output = await cacheItem.getPromise();

				if (!output) {
					output = new RawSource(templatePreview(spritesList), false);

					await cacheItem.storePromise(output);
				}

				compilation.emitAsset('sprites-preview.html', output);
			}
		}
	}

	/**
	 * Generate the SVG sprite with Svgstore
	 * @param {Array<Svgs>} svgsOptimized SVGs list
	 * @returns {String} Sprite string
	 */
	generateSprite(svgsOptimized) {
		const sprites = svgstore(this.options.svgstoreConfig);

		svgsOptimized.forEach((svg) => {
			sprites.add(svg.name, svg.content);
		});

		return sprites.toString();
	}
	/**
	 * Get the filename for the asset compilation
	 * Placeholder [name], [hash], [chunkhash], [content] are automatically replaced
	 * @param {String} entryName Entrypoint name
	 * @param {String} output Sprite content
	 * @returns {String} Sprite filename
	 */
	getFileName({ entryName, output }) {
		let filename = this.options.filename;
		// Check if the filename option contains the placeholder [name]
		// [name] corresponds to the entrypoint name
		if (REGEXP_NAME.test(filename)) {
			filename = filename.replace('[name]', entryName);
		}
		// Check if the filename option contains the placeholder [hash]
		// [hash] corresponds to the Webpack compilation hash
		if (REGEXP_HASH.test(filename)) {
			filename = filename.replace('[hash]', this.getBuildHash());
		}
		// Check if the filename option contains the placeholder [chunkhash]
		// [chunkhash] corresponds to the hash of the chunk content
		if (REGEXP_CHUNKHASH.test(filename)) {
			filename = filename.replace('[chunkhash]', this.getChunkHash(entryName));
		}
		// Check if the filename option contains the placeholder [contenthash]
		// [contenthash] corresponds to the sprite content hash
		if (REGEXP_CONTENTHASH.test(filename)) {
			filename = filename.replace('[contenthash]', this.getContentHash(output));
		}
		return filename;
	}
	/**
	 * Get the compilation hash
	 * @returns {String} Compilation hash
	 */
	getBuildHash() {
		return this.compilation.hash;
	}
	/**
	 * Get the chunk hash according to the entrypoint content
	 * @returns {String} Chunk hash
	 */
	getChunkHash(entryName) {
		const chunks = this.compilation.entrypoints.get(entryName).chunks;
		return chunks.length ? chunks[0].hash : '';
	}
	/**
	 * Get the contenthash according to the sprite content
	 * @returns {String} Sprite content hash
	 */
	getContentHash(output) {
		const { hashFunction, hashDigest } = this.compilation.outputOptions;
		return util.createHash(hashFunction).update(output).digest(hashDigest);
	}
}
// @ts-ignore
SvgSprite.loader = require.resolve('./loader');
module.exports = SvgSprite;
```

**Reply by alexander-akait** [reply]
On my computer without cache, compilation takes around 900ms-1s, with cache it is 100-120ms (if you didn't change svg files), also this improve perf for watch/serve mode

**Reply by yoriiis** [reply]
@alexander-akait Thanks for the changes and for the optimizations. I regularly work with _class properties_ and was unaware of potential memory leaks, thanks for the info.

I have some problems with your version:

- During the second build, the sprite is empty
- There are errors with using the `reduce` function, an initial value is missing but I don't know which one to provide. The error message: `TypeError: reducing an empty array with no initial value`
- In the sort function the field `NormalModule.name` does not seem to be available
- Is access to private property `NormalModule._source` correct?

---

I've added your changes on [svg-chunk-webpack-plugin#8](https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8) if you want to test.

Also, would you like me to add you as a collaborator on the project?

**Reply by alexander-akait** [reply]
>  have some problems with your version:

Weird, I have tested it locally, but I write it fast, so can make some typos, yeah, you can invite me and we will finish it, because logic is written, most of issues are likely just typos

**Reply by alexander-akait** [reply]
And how you run tests?

**Reply by alexander-akait** [reply]
> Is access to private property NormalModule._source correct?

And yes and no - we should write another logic to collect sources, but it requires more investigation, so I think we can safely use it until we will have problems with it (mostly never)

**Reply by yoriiis** [reply]
@alexander-akait no problem. I've sent you an invitation to join the project.

The project doesn't have a contribution guide yet, but quickly, you can use these commands:

```bash
# Install Node.js dependencies
nvm use
npm install

# Start TS watcher
npm run dev:types

# Use example (for the test)
npm run start:example
npm run build:example
```

Then, you can check the [dist](https://github.com/yoriiis/svg-chunk-webpack-plugin/tree/master/example/dist) directory for generated files.

**Comment by yoriiis** [comment]
The plugin has been rewritten (https://github.com/yoriiis/svg-chunk-webpack-plugin/pull/8) to be compatible with the webpack cache. A big thank you @alexander-akait for your help and your availability!

**Reply by alexander-akait** [reply]
No problem, you are welcome
